iroh-gateway/src/bad_bits.rs

```diff
@@ -0,0 +1,206 @@
+use crypto::{digest::Digest, sha2::Sha256};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, sync::Arc, time::Duration};
+use tokio::{sync::RwLock, task::JoinHandle};
+use tracing::log::error;
+
+const BAD_BITS_UPDATE_INTERVAL: Duration = Duration::from_secs(3600 * 8);
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct BadBitsAnchor {
+    pub anchor: String,
+}
+
+#[derive(Debug)]
+pub struct BadBits {
+    pub last_updated: time::Instant,
+    pub denylist: HashMap<String, bool>,
+}
+
+impl BadBits {
+    pub fn new() -> Self {
+        Self {
+            last_updated: time::Instant::now(),
+            denylist: HashMap::new(),
+        }
+    }
+
+    pub fn update(&mut self, denylist: HashMap<String, bool>) {
+        self.last_updated = time::Instant::now();
+        self.denylist = denylist;
+    }
+
+    pub fn is_bad(&self, cid: &str, path: &str) -> bool {
+        let hash = BadBits::to_anchor(cid, path);
+        *self.denylist.get(&hash).unwrap_or(&false)
+    }
+
+    pub fn to_anchor(cid: &str, path: &str) -> String {
+        let mut hasher = Sha256::new();
+        hasher.input_str(cid);
+        if !path.is_empty() {
+            if path.starts_with('/') {
+                hasher.input_str(path);
+            } else {
+                hasher.input_str("/");
+                hasher.input_str(path);
+            }
+        }
+        hasher.result_str()
+    }
+}
+
+impl Default for BadBits {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub fn bad_bits_update_handler(bad_bits: Arc<RwLock<BadBits>>) -> JoinHandle<()> {
+    tokio::spawn(async move {
+        loop {
+            let denylist_uri = "http://badbits.dwebops.pub/denylist.json";
+            let res = reqwest::get(denylist_uri).await.unwrap();
+            if res.status().is_success() {
+                let body = res.bytes().await.unwrap();
+                let body = serde_json::from_slice::<Vec<BadBitsAnchor>>(&body[..]).unwrap();
```

Comment:
you can use `https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.json` directly, avoiding the manual call to `bytes()`